File: 8-kyu-kata/welcome-to-the-city/python/welcome_to_the_city_test.py
# 8 kyu - Welcome to the City  [ ID: 5302d846be2a9189af0001e4  (welcome-to-the-city) ]
# URL: https://www.codewars.com/kata/5302d846be2a9189af0001e4
# Category: REFERENCE  |  Tags: STRINGS | FUNDAMENTALS
# ******************************************************************************
import codewars_test as test
from solution import say_hello


@test.describe("Fixed Tests")
def fixed_tests():
    @test.it("Basic Test Cases")
    def basic_test_cases():
        test.assert_equals(
            say_hello(["John", "Smith"], "Phoenix", "Arizona"),
            "Hello, John Smith! Welcome to Phoenix, Arizona!",
            "Hello, John Smith! Welcome to Phoenix, Arizona!",
        )
        test.assert_equals(
            say_hello(["Franklin", "Delano", "Roosevelt"], "Chicago", "Illinois"),
            "Hello, Franklin Delano Roosevelt! Welcome to Chicago, Illinois!",
        )
        test.assert_equals(
            say_hello(["Wallace", "Russel", "Osbourne"], "Albany", "New York"),
            "Hello, Wallace Russel Osbourne! Welcome to Albany, New York!",
        )
        test.assert_equals(
            say_hello(["Lupin", "the", "Third"], "Los Angeles", "California"),
            "Hello, Lupin the Third! Welcome to Los Angeles, California!",
        )
        test.assert_equals(
            say_hello(["Marlo", "Stanfield"], "Baltimore", "Maryland"),
            "Hello, Marlo Stanfield! Welcome to Baltimore, Maryland!",
        )


@test.describe("Random Tests")
def random_tests():

    from random import randint

    names = [
        ["Frank", "Underwood"],
        ["Claire", "Underwood"],
        ["Douglas", '"Doug"', "Stamper"],
        ["Zoe", "Barnes"],
        ["Peter", "Russo"],
        ["Remy", "Danton"],
        ["Former", "President", "Garrett", "Walker"],
    ]
    cities = ["New York City", "Los Angeles", "Chicago", "Houston", "Philadelphia", "Detroit"]
    states = ["New York", "California", "Illinois", "Texas", "Pennsylvania", "Michigan"]

    sol = lambda name, city, state: "Hello, " + " ".join(name) + "! Welcome to " + city + ", " + state + "!"

    for _ in range(40):
        name = names[randint(0, len(names) - 1)]
        location = randint(0, len(cities) - 1)
        city = cities[location]
        state = states[location]
        expected = sol(name, city, state)

        @test.it(f'Testing for {" ".join(name)} in {city}, {state}')
        def test_case():
            test.assert_equals(say_hello(name, city, state), expected)

File: 8-kyu-kata/welcome-to-the-city/javascript/welcomeToTheCity.Test.js
// 8 kyu - Welcome to the City  [ ID: 5302d846be2a9189af0001e4  (welcome-to-the-city) ]
// URL: https://www.codewars.com/kata/5302d846be2a9189af0001e4
// Category: REFERENCE  |  Tags: STRINGS | FUNDAMENTALS
// *****************************************************************************
const chai = require("chai");
const assert = chai.assert;
chai.config.truncateThreshold=0;

describe("Basic Tests", () => {
  it("Should pass Basic Tests", () => {
    assert.strictEqual(sayHello(['John', 'Smith'], 'Phoenix', 'Arizona'), 'Hello, John Smith! Welcome to Phoenix, Arizona!')
    assert.strictEqual(sayHello(['Franklin','Delano','Roosevelt'], 'Chicago', 'Illinois'), 'Hello, Franklin Delano Roosevelt! Welcome to Chicago, Illinois!')
    assert.strictEqual(sayHello(['Wallace','Russel','Osbourne'],'Albany','New York'), 'Hello, Wallace Russel Osbourne! Welcome to Albany, New York!')
  });
});

describe("Random Tests", () => {
  
  function sayHelloSolution( name, city, state ) {
    return 'Hello, ' + name.join(' ') + '! Welcome to ' + city + ', ' + state + '!';
  }
  
  const names=[["Frank", "Underwood"],["Claire", "Underwood"],["Douglas", '"Doug"', "Stamper"],["Zoe", "Barnes"], ["Peter", "Russo"], ["Remy", "Danton"], ["Former", "President", "Garrett", "Walker"]],
        cities=["New York City","Los Angeles","Chicago","Houston","Philadelphia", "Detroit"],
        states=["New York","California","Illinois","Texas","Pennsylvania", "Michigan"];
      
  for (let i = 0; i < 40; i++) {
    let name = names[Math.floor(Math.random() * names.length)],
        location = Math.floor(Math.random() * cities.length),
        city = cities[location],
        state = states[location],
        expected = sayHelloSolution( name, city, state );
    it(`Testing for ["${name.join("\", \"")}"], "${city}" and "${state}"`, () => {
      assert.strictEqual(sayHello(name, city, state), expected);
    });    
  }  
  
});


